### PR TITLE
[Broadcaster] Add support for multiple inputs

### PR DESCRIPTION
**Pull request by sgfn**

### CODE REVIEW

**Review by mickel8 — COMMENTED** [review]
I feel like we are mixing terms - stream, input and event. E.g. we have stream channel where we emit stream_added and stream_removed events. In Forwarder, sometimes we use `stream` term, and some other time `input` term. 

I think, I would model this as:

* we have a single **Event** 
* this event can have multiple **inputs** and multiple **outputs**

I would resign from using stream term.

Alternatively, use **publishers** and **subscribers**

Any thoughts? :thinking:

File: broadcaster/lib/broadcaster_web/controllers/media_controller.ex [hunk 1]
```diff
@@ -72,15 +82,15 @@ defmodule BroadcasterWeb.MediaController do
...
-      {:ok, _pid} ->
-        conn
-        |> put_resp_header("content-type", "text/event-stream")
-        |> put_resp_header("connection", "keep-alive")
-        |> put_resp_header("cache-control", "no-cache")
-        |> send_chunked(200)
-        |> update_layers()
-
+    with {:ok, _pid} <- PeerSupervisor.fetch_pid(resource_id),
+         [stream_id, _pc_id] <- String.split(resource_id, "-") do
```
Comment (mickel8) [inline broadcaster/lib/broadcaster_web/controllers/media_controller.ex:86]: To be fully correct, `resource_id` is in fact `pc_id` so after `String.splti` we get `stream_id` and some postifx. or at least that's how I understand it looking at line 43 where as `resource_id`, we return `pc_id`, which in turn has embeded `stream_id`. That might sound like a nitpick but for someone who was not writing the code, it is a bit missleading :stuck_out_tongue:

File: broadcaster/assets/js/home.js [hunk 2]
```diff
@@ -1,56 +1,110 @@
...
-async function sendCandidate(candidate) {
-  const response = await fetch(patchEndpoint, {
-    method: 'PATCH',
-    cache: 'no-cache',
-    headers: {
-      'Content-Type': 'application/trickle-ice-sdpfrag',
-    },
-    body: candidate,
+async function connectMedia(socket) {
+  const channel = socket.channel('stream:signalling');
```
[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/assets/js/home.js:19]: https://dictionary.cambridge.org/dictionary/english/signalling 🇬🇧 🇬🇧

File: broadcaster/lib/broadcaster_web/controllers/page_html/panel.html.heex [hunk 3]
```diff
@@ -120,8 +120,7 @@
       </button>
     </div>
     <div id="videoplayer-wrapper" class="flex flex-1 flex-col min-h-0 gap-4">
-      <video id="previewPlayer" class="m-auto rounded-lg bg-black h-full" autoplay controls muted>
```
[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/lib/broadcaster_web/controllers/page_html/panel.html.heex:123]: Why did we need controls for the preview player in the first place? It's supposed to show you what's being transmitted, what's the use of pausing/rewinding it?

File: broadcaster/lib/broadcaster_web/controllers/page_html/home.html.heex [hunk 4]
```diff
@@ -1,13 +1,10 @@
...
-  <div id="videoplayer-wrapper" class="flex flex-1 flex-col gap-4 overflow-y-scroll">
-    <video
-      id="videoplayer"
-      class="rounded-xl max-h-[1080px] w-full bg-black"
-      autoplay
-      controls
-      muted
-    >
-    </video>
+  <div id="videoplayer-wrapper-wrapper" class="w-full h-full gap-2 p-2">
```
[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/lib/broadcaster_web/controllers/page_html/home.html.heex:2]: And here I thought I'd get bonus points for wrapping a wrapper...

**Comment by sgfn** [comment]
> Alternatively, use **publishers** and **subscribers**

will go with that, don't see how a single `Event` would be supposed to work in this context

File: broadcaster/lib/broadcaster/forwarder.ex [hunk 5]
```diff
@@ -68,177 +77,195 @@ defmodule Broadcaster.Forwarder do
...
+  def handle_info({:ex_webrtc, pc, {:connection_state_change, :connected}}, state)
+      when is_map_key(state.pending_outputs, pc) do
+    {input_id, state} = pop_in(state, [:pending_outputs, pc])
+
     state =
-      if Enum.member?(state.pending_outputs, pc) do
-        pending_outputs = List.delete(state.pending_outputs, pc)
-        {audio_track, video_track} = get_tracks(pc, :sender)
+      with {:ok, input_id} <- expand_default_input_id(input_id, state),
+           {:ok, _input_pc, input} <- find_input(input_id, state) do
```
[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/lib/broadcaster/forwarder.ex:155]: I recall there were reasons why I decided to implement things this way, just can't remember what these reasons were precisely

Will try to revert to the basic/simpler version nonetheless

File: broadcaster/assets/js/home.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mickel8) [reply broadcaster/assets/js/home.js:19]: WebRTC is different :D 

https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Signaling_and_video_calling

**Comment by mickel8** [comment]
> will go with that, don't see how a single Event would be supposed to work in this context

I would use single **Event** as top entity to e.g. name a div or to name a channel e.g. EventChannel where we have stream_added/publisher_added

File: broadcaster/lib/broadcaster_web/controllers/page_html/home.html.heex (hunk 4 above)
[the remark this answers is not in the document]
Comment (mickel8) [reply broadcaster/lib/broadcaster_web/controllers/page_html/home.html.heex:2]: Sorry, my bad

File: broadcaster/lib/broadcaster_web/controllers/page_html/panel.html.heex (hunk 3 above)
[the remark this answers is not in the document]
Comment (mickel8) [reply broadcaster/lib/broadcaster_web/controllers/page_html/panel.html.heex:123]: I think this was for muting

[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/lib/broadcaster_web/controllers/page_html/panel.html.heex:123]: Still seems weird to me (what would incline someone to unmute the muted-by-default preview player?) but OK, confirm you want me to return the controls and I'll do it

File: broadcaster/assets/js/home.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/assets/js/home.js:19]: Screenshot from the very page you linked...

image

[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/assets/js/home.js:19]: Confirm you want to force the american spelling?

File: broadcaster/assets/js/home.js [hunk 6]
```diff
@@ -1,56 +1,110 @@
...
+      );
+
+      statusMessage.innerText =
+        'Unable to join the stream, try again in a few minutes';
+      statusMessage.classList.remove('hidden');
+    });
 }
 
-async function connectMedia() {
+async function connectStream(streamId) {
```
[the remark this answers is not in the document]
Comment (sgfn) [reply broadcaster/assets/js/home.js:61]: What exactly would you like to see extracted to this WHEPClient?

File: broadcaster/assets/js/whep-client.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
+const pcConfig = { iceServers: [{ urls: 'stun:stun.l.google.com:19302' }] };
+
+export class WHEPClient {
```
Comment (mickel8) [inline broadcaster/assets/js/whep-client.js:3]: Nice!! :+1:

**Review by mickel8 — APPROVED** [review]

**Pull request merged**